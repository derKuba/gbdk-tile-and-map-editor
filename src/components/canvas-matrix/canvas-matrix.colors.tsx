export enum Colors {
    WHITE = 0,
    GREY = 1,
    GREY_DARK = 2,
    BLACK = 3,
}
