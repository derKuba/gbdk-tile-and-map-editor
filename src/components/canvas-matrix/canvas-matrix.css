:host {
  display: block;
}

.canvas-matrix-field-old {
  display: grid;
  grid-template-columns: repeat(8, 62px);
  grid-gap: 0px;
  border: 1px solid black;
  border-bottom: 0;
  width: 495px;
}

.canvas-matrix-field{
  display: flex;
  border: 1px solid black;
  flex-direction: column;
}

.matrixRow {
  display: inline-flex;
}

.matrixRow span {
  font-size: 12px;
  margin-left: 3px;
}