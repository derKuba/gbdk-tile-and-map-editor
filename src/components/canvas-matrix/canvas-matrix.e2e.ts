import { newE2EPage } from '@stencil/core/testing';

describe('mvjp-canvas-matrix', () => {
    it('renders', async () => {
        const page = await newE2EPage();
        await page.setContent('<mvjp-canvas-matrix></mvjp-canvas-matrix>');

        const element = await page.find('mvjp-canvas-matrix');
        expect(element).toHaveClass('hydrated');
    });
});
