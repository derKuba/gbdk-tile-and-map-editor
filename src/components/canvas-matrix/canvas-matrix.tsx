import { Component, Host, h, Prop } from '@stencil/core';

import { PaintModus } from '../../models/paintModus';
import { paintBucket } from '../../utils/paintBucketService';
import { Settings } from '../../models/settingsModel';

import Tunnel from '../../provider/StateProvider';

@Component({
    tag: 'mvjp-canvas-matrix',
    styleUrl: 'canvas-matrix.css',
    shadow: true,
})
export class CanvasMatrix {
    private isMoving: boolean = false;

    /**
     *   choosen colorcode
     */
    @Prop() choosenColour: number;

    /**
     *   state of the entire matrix
     */
    @Prop() matrix: number[][][];

    /**
     *   active sprite tile
     */
    @Prop() indexTileMatrix: number;

    /**
     *   set matrix
     */
    @Prop() setMatrix: Function;

    /**
     *   brush or bucket
     */
    @Prop() paintModus: PaintModus;

    /**
     *   application settings
     */
    @Prop() settings: Settings;

    private clickMatrixTile = position => {
        if (this.paintModus === PaintModus.BRUSH) {
            this.matrix[this.indexTileMatrix][position.y][position.x] = this.choosenColour;
        } else if (this.paintModus === PaintModus.BUCKET) {
            if (this.matrix[this.indexTileMatrix][position.y][position.x] === this.choosenColour) return;

            this.matrix[this.indexTileMatrix] = paintBucket(
                this.matrix[this.indexTileMatrix],
                position,
                this.matrix[this.indexTileMatrix][position.y][position.x],
                this.choosenColour
            );
        }

        this.matrix = [...this.matrix];
        this.setMatrix(this.matrix);
    };

    private mouseUp = () => {
        this.isMoving = false;
    };

    private mouseDown = () => {
        this.isMoving = true;
    };

    private mouseMove = event => {
        if (this.isMoving && this.paintModus === PaintModus.BRUSH) {
            try {
                const targetPos = JSON.parse(event.target.dataset.pos);
                this.clickMatrixTile(targetPos);
            } catch (jsonParseTargetError) {
                console.log(jsonParseTargetError); // eslint-disable-line no-undef
            }
        }
    };

    render() {
        const tileMatrix = this.matrix[this.indexTileMatrix];

        const field = tileMatrix.map((tileRow: number[], y: number) => (
            <div class="matrixRow" key={`tile_at_${y}`}>
                {tileRow.map((colorCode: number, x: number) => (
                    <mvjp-canvas-matrix-tile
                        key={`tile_at_${y}_${x}`}
                        position={{ y, x }}
                        handleClick={this.clickMatrixTile}
                        colour={colorCode}
                        showGrid={this.settings.isGridVisible}>
                        {this.settings.isFieldNumberVisible ? <span>{`${y}_${x}`}</span> : null}
                    </mvjp-canvas-matrix-tile>
                ))}
            </div>
        ));

        return (
            <Host>
                <div
                    class="canvas-matrix-field"
                    onMouseMove={this.mouseMove}
                    onMouseDown={this.mouseDown}
                    onMouseUp={this.mouseUp}>
                    {field}
                </div>
            </Host>
        );
    }
}

Tunnel.injectProps(CanvasMatrix, ['choosenColour', 'matrix', 'setMatrix', 'indexTileMatrix', 'paintModus', 'settings']); // eslint-disable-line @stencil/ban-side-effects
