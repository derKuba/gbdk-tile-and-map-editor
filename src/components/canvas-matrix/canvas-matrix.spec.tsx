import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { CanvasMatrix } from './canvas-matrix';

describe('mvjp-canvas-matrix', () => {
    let matrix: number[][][] = [];
    beforeEach(() => {
        for (let i = 0; i < 12; i++) {
            matrix.push(Array.from({ length: 8 }, () => Array.from({ length: 8 }, () => 0)));
        }

        matrix[0] = [
            [1, 1, 1, 0, 1, 1, 1, 0],
            [0, 0, 1, 0, 1, 0, 0, 1],
            [0, 0, 1, 0, 1, 1, 1, 0],
            [0, 0, 1, 0, 1, 0, 0, 0],
            [0, 1, 1, 0, 1, 0, 0, 0],
            [1, 0, 1, 0, 1, 0, 0, 0],
            [0, 1, 0, 0, 1, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
        ];
    });

    it('builds', () => {
        expect(new CanvasMatrix()).toBeTruthy();
    });

    it('matches snapshot', async () => {
        const choosenColour = 1;
        const indexTileMatrix = 0;

        const settings = {
            fieldWidth: 8,
            fieldHeight: 8,
            spriteAmount: 12,
            isGridVisible: true,
            isFieldNumberVisible: true,
        };

        const page = await newSpecPage({
            components: [CanvasMatrix],
            template: () => {
                return (
                    <mvjp-canvas-matrix
                        choosenColour={choosenColour}
                        matrix={matrix}
                        settings={settings}
                        paintModus={0}
                        indexTileMatrix={indexTileMatrix}
                        setMatrix={() => {}}></mvjp-canvas-matrix>
                );
            },
        });

        expect(page.root).toMatchSnapshot();
    });

    it('should change color in matrix', () => {
        const canvasMatrix = new CanvasMatrix();
        canvasMatrix.setMatrix = () => {};
        canvasMatrix.matrix = matrix;
        canvasMatrix.choosenColour = 2;
        canvasMatrix.indexTileMatrix = 0;
        canvasMatrix.paintModus = 0;

        canvasMatrix.clickMatrixTile({ x: 0, y: 0 });
        expect(canvasMatrix.matrix[0][0][0]).toBe(2);
    });

    it('should change mousemoving', () => {
        const canvasMatrix = new CanvasMatrix();
        canvasMatrix.setMatrix = () => {};
        canvasMatrix.matrix = matrix;
        canvasMatrix.choosenColour = 2;
        canvasMatrix.indexTileMatrix = 0;
        canvasMatrix.paintModus = 0;

        canvasMatrix.clickMatrixTile = jest.fn();
        canvasMatrix.mouseUp();

        const event = {
            target: {
                dataset: { pos: '{"y":1,"x":6}' },
            },
        };
        canvasMatrix.mouseMove(event);
        expect(canvasMatrix.clickMatrixTile).not.toBeCalled();

        canvasMatrix.mouseDown();
        canvasMatrix.mouseMove(event);
        expect(canvasMatrix.clickMatrixTile).toBeCalledWith({ x: 6, y: 1 });
    });

    it('should throw json parse error', () => {
        const canvasMatrix = new CanvasMatrix();
        canvasMatrix.setMatrix = () => {};
        canvasMatrix.matrix = matrix;
        canvasMatrix.choosenColour = 2;
        canvasMatrix.indexTileMatrix = 0;
        canvasMatrix.paintModus = 0;

        canvasMatrix.clickMatrixTile = jest.fn();
        canvasMatrix.mouseDown();

        const event = {
            target: {
                dataset: { pos: '{ y: 1, x: 6 }' },
            },
        };
        canvasMatrix.mouseMove(event);
        expect(canvasMatrix.clickMatrixTile).not.toBeCalled();
    });

    it('should change color in matrix', () => {
        const canvasMatrix = new CanvasMatrix();
        canvasMatrix.setMatrix = () => {};
        canvasMatrix.matrix = matrix;
        canvasMatrix.choosenColour = 2;
        canvasMatrix.indexTileMatrix = 0;
        canvasMatrix.paintModus = 1;

        canvasMatrix.clickMatrixTile({ x: 0, y: 0 });
        expect(canvasMatrix.matrix[0]).toEqual([
            [2, 2, 2, 0, 1, 1, 1, 0],
            [0, 0, 2, 0, 1, 0, 0, 1],
            [0, 0, 2, 0, 1, 1, 1, 0],
            [0, 0, 2, 0, 1, 0, 0, 0],
            [0, 2, 2, 0, 1, 0, 0, 0],
            [1, 0, 2, 0, 1, 0, 0, 0],
            [0, 1, 0, 0, 1, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
        ]);

        canvasMatrix.clickMatrixTile({ x: 0, y: 0 });
    });
});
