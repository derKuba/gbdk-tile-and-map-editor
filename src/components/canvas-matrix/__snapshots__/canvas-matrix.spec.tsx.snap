// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mvjp-canvas-matrix matches snapshot 1`] = `
<mvjp-canvas-matrix>
  <mock:shadow-root>
    <div class="canvas-matrix-field">
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            0_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            0_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            0_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            1_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            1_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            1_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            1_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            1_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            1_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            1_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            1_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            2_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            2_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            2_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            2_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            2_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            2_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            2_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            2_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            3_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            3_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            3_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            4_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            4_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            4_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            4_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            4_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            4_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            4_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            4_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            5_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            5_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            5_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            5_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            5_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            5_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            5_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            5_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            6_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="1" showgrid="">
          <span>
            6_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            6_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
      <div class="matrixRow">
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_0
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_1
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_2
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_3
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_4
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_5
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_6
          </span>
        </mvjp-canvas-matrix-tile>
        <mvjp-canvas-matrix-tile colour="0" showgrid="">
          <span>
            7_7
          </span>
        </mvjp-canvas-matrix-tile>
      </div>
    </div>
  </mock:shadow-root>
</mvjp-canvas-matrix>
`;
